window-manager/src/main/java/com/codelab/foldables/window_manager/MainActivity.kt

```diff
@@ -20,59 +20,63 @@ package com.codelab.foldables.window_manager
 
 import android.graphics.Rect
 import android.os.Bundle
-import android.os.Handler
-import android.os.Looper
 import android.util.TypedValue
 import android.view.View
 import androidx.appcompat.app.AppCompatActivity
 import androidx.constraintlayout.widget.ConstraintSet
-import androidx.core.util.Consumer
-import androidx.window.WindowLayoutInfo
-import androidx.window.WindowManager
+import androidx.window.layout.FoldingFeature
+import androidx.window.layout.WindowInfoRepository
+import androidx.window.layout.WindowInfoRepository.Companion.windowInfoRepository
+import androidx.window.layout.WindowLayoutInfo
+import androidx.window.layout.WindowMetricsCalculator
 import com.codelab.foldables.window_manager.databinding.ActivityMainBinding
-import java.util.concurrent.Executor
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.MainScope
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
 
 class MainActivity : AppCompatActivity() {
-
-    private lateinit var wm: WindowManager
-    private val layoutStateChangeCallback = LayoutStateChangeCallback()
     private lateinit var binding: ActivityMainBinding
 
-    private fun runOnUiThreadExecutor(): Executor {
-        val handler = Handler(Looper.getMainLooper())
-        return Executor() {
-            handler.post(it)
-        }
-    }
+    private lateinit var windowInfoRepository: WindowInfoRepository
+    private val scope = MainScope()
 
+    @ExperimentalCoroutinesApi
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         binding = ActivityMainBinding.inflate(layoutInflater)
         setContentView(binding.root)
 
-        wm = WindowManager(this)
-    }
+        windowInfoRepository = windowInfoRepository()
 
-    override fun onAttachedToWindow() {
-        super.onAttachedToWindow()
-        wm.registerLayoutChangeCallback(
-            runOnUiThreadExecutor(),
-            layoutStateChangeCallback
-        )
+        obtainWindowMetrics()
+        onWindowLayoutInfoChange(windowInfoRepository)
     }
 
-    override fun onDetachedFromWindow() {
-        super.onDetachedFromWindow()
-        wm.unregisterLayoutChangeCallback(layoutStateChangeCallback)
+    override fun onDestroy() {
+        super.onDestroy()
+        scope.cancel()
     }
 
-    private fun printLayoutStateChange(newLayoutInfo: WindowLayoutInfo) {
+    private fun obtainWindowMetrics() {
+        val wmc = WindowMetricsCalculator.getOrCreate()
         binding.windowMetrics.text =
-            "CurrentWindowMetrics: ${wm.currentWindowMetrics.bounds.flattenToString()}\n" +
-                "MaximumWindowMetrics: ${wm.maximumWindowMetrics.bounds.flattenToString()}"
+            "CurrentWindowMetrics: ${wmc.computeCurrentWindowMetrics(this).bounds.flattenToString()}\n" +
```

Comment:
nit: these lines are longer than 100 chars that is the stated max in the Android style guide.

Reply:
Addressed in new commits. Thanks Pietro!